crates/circuit/src/dag_circuit.rs

```diff
@@ -7095,6 +6965,273 @@ impl DAGCircuit {
+    /// recorded nodes.
+    pub fn end(mut self) -> DAGCircuit {
+        // Re-connects all of the output nodes with their respective last nodes.
+        // Add the output_nodes back to qargs
+        for (qubit, node) in self
+            .last_qubits
+            .into_iter()
+            .enumerate()
+            .filter_map(|(qubit, node)| node.map(|node| (qubit, node)))
+        {
+            let output_node = self.dag.qubit_io_map[qubit][1];
+            self.dag
+                .dag
+                .add_edge(node, output_node, Wire::Qubit(Qubit(qubit as u32)));
+        }
+
+        // Add the output_nodes back to cargs
+        for (clbit, node) in self
+            .last_clbits
+            .into_iter()
+            .enumerate()
+            .filter_map(|(clbit, node)| node.map(|node| (clbit, node)))
+        {
+            let output_node = self.dag.clbit_io_map[clbit][1];
+            self.dag
+                .dag
+                .add_edge(node, output_node, Wire::Clbit(Clbit(clbit as u32)));
+        }
+
+        // Add the output_nodes back to vars
+        for (var, node) in self
+            .last_vars
+            .into_iter()
+            .enumerate()
+            .filter_map(|(var, node)| node.map(|node| (var, node)))
+        {
+            let output_node = self.dag.var_io_map[var][1];
+            self.dag
+                .dag
+                .add_edge(node, output_node, Wire::Var(Var(var as u32)));
+        }
+        self.dag
+    }
+
+    /// Applies a new operation to the back of the circuit. This variant works with non-owned bit indices.
+    pub fn apply_operation_back(
+        &mut self,
+        py: Python,
+        op: PackedOperation,
+        qubits: Option<Cow<[Qubit]>>,
+        clbits: Option<Cow<[Clbit]>>,
+        params: Option<Box<SmallVec<[Param; 3]>>>,
+        label: Option<String>,
+        #[cfg(feature = "cache_pygates")] py_op: Option<PyObject>,
+    ) -> NodeIndex {
+        let instruction = self.pack_instruction(
+            op,
+            qubits,
+            clbits,
+            params,
+            label,
+            #[cfg(feature = "cache_pygates")]
+            py_op,
+        );
+        self.push_instruction_back(py, instruction)
+    }
+
+    /// Pushes a valid [PackedInstruction] to the back ot the circuit.
+    pub fn push_instruction_back(&mut self, py: Python<'_>, inst: PackedInstruction) -> NodeIndex {
+        let op_name = inst.op.name();
+        let (all_cbits, vars): (Vec<Clbit>, Option<Vec<Var>>) = {
+            if self.dag.may_have_additional_wires(py, &inst) {
+                let mut clbits: HashSet<Clbit> =
+                    HashSet::from_iter(self.dag.cargs_interner.get(inst.clbits).iter().copied());
+                let (additional_clbits, additional_vars) = self
+                    .dag
+                    .additional_wires(py, inst.op.view())
+                    .expect("Error trying to find additional wires for instruction");
```

Comment:
Is there a reason not to return a `PyResult` from this method?

Reply:
There isn't a reason currently, adding...